File: src/main/scala/hod/javacatchup/Euler72.java
package hod.javacatchup;

import hod.EulerUtils;

import java.util.concurrent.atomic.AtomicLong;
import java.util.stream.IntStream;

public class Euler72 {

    private static final AtomicLong counter = new AtomicLong();

    public static void main(String[] args) {
        var limit = 8;
        var limit2 = 1000000;
        var max = limit2;
        var ds = numberIterator(1, max);
        var solution = ds.parallel()
                .mapToLong(d -> {
                               counter.incrementAndGet();
                               if (counter.get() % 1000 == 0) {
                                   System.out.println(counter.get());
                               }
                               return numberIterator(1, d - 1)
                                       .filter(n -> EulerUtils.isReducedProperFractionJ(n, d))
                                       .count();
                           }
                ).reduce(Long::sum);
        System.out.println("solution = " + solution);
    }

    private static IntStream numberIterator(final int from, final int to) {
        return IntStream.range(from, to + 1);
    }
}

File: src/main/scala/hod/euler/Euler72.scala
package hod.euler

import hod.EulerUtils

import java.util.concurrent.atomic.AtomicInteger
import scala.collection.parallel.CollectionConverters._

object Euler72 {

  def main(args: Array[String]): Unit = {
    measured {
      val max = 1000000
      val count = new AtomicInteger()
      val solution = (1 to max).par.map { n =>
        if (count.incrementAndGet() % 1000 == 0) {
          print(".")
        }
        (1 until n).count { d =>
          gcdEuclid(d, n) == 1
        }
      }.map(_.toLong).sum
      println("\n" + solution)
    }
  }

}
